File: src/go/helpers/data-structures/trie.go
package datastructures

import "sort"

type TrieNode struct {
	children    map[rune]*TrieNode
	isEndOfWord bool
}

type Trie struct {
	root *TrieNode
}

func NewTrie() *Trie {
	return &Trie{
		root: NewTrieNode(),
	}
}

func NewTrieNode() *TrieNode {
	return &TrieNode{
		children: make(map[rune]*TrieNode),
	}
}

func (trie *Trie) Insert(word string) {
	currentNode := trie.root

	for _, character := range word {
		nextNode, exists := currentNode.children[character]
		if !exists {
			nextNode = NewTrieNode()
			currentNode.children[character] = nextNode
		}

		currentNode = nextNode
	}

	currentNode.isEndOfWord = true
}

func (trie *Trie) Autocomplete(prefix string) []string {
	prefixNode := trie.findNode(prefix)
	if prefixNode == nil {
		return []string{}
	}

	results := make([]string, 0)

	prefixRunes := []rune(prefix)
	trie.collectWords(prefixNode, prefixRunes, &results, false)

	return results
}

func (trie *Trie) CountWords() int {
	return trie.countWordsRecursive(trie.root)
}

func (trie *Trie) GetAllWords() []string {
	words := make([]string, 0)
	trie.collectWords(trie.root, []rune{}, &words, true)
	return words
}

func (trie *Trie) Traverse(visit func(word string)) {
	trie.traverseRecursive(trie.root, []rune{}, visit)
}

func (trie *Trie) findNode(prefix string) *TrieNode {
	currentNode := trie.root

	for _, character := range prefix {
		nextNode, exists := currentNode.children[character]
		if !exists {
			return nil
		}

		currentNode = nextNode
	}

	return currentNode
}

func (trie *Trie) countWordsRecursive(currentNode *TrieNode) int {
	if currentNode == nil {
		return 0
	}

	count := 0
	if currentNode.isEndOfWord {
		count = 1
	}

	for _, childNode := range currentNode.children {
		count += trie.countWordsRecursive(childNode)
	}

	return count
}

func (trie *Trie) collectWords(
	currentNode *TrieNode,
	prefixRunes []rune,
	results *[]string,
	sortChildren bool,
) {
	if currentNode == nil {
		return
	}

	if currentNode.isEndOfWord {
		*results = append(*results, string(prefixRunes))
	}

	childKeys := trie.childKeys(currentNode, sortChildren)

	for _, character := range childKeys {
		childNode := currentNode.children[character]

		prefixRunes = append(prefixRunes, character)
		trie.collectWords(childNode, prefixRunes, results, sortChildren)
		prefixRunes = prefixRunes[:len(prefixRunes)-1]
	}
}

func (trie *Trie) traverseRecursive(
	currentNode *TrieNode,
	prefixRunes []rune,
	visit func(word string),
) {
	if currentNode == nil {
		return
	}

	if currentNode.isEndOfWord {
		visit(string(prefixRunes))
	}

	childKeys := trie.childKeys(currentNode, true)

	for _, character := range childKeys {
		childNode := currentNode.children[character]

		prefixRunes = append(prefixRunes, character)
		trie.traverseRecursive(childNode, prefixRunes, visit)
		prefixRunes = prefixRunes[:len(prefixRunes)-1]
	}
}

func (trie *Trie) childKeys(currentNode *TrieNode, sortKeys bool) []rune {
	keys := make([]rune, 0, len(currentNode.children))

	for character := range currentNode.children {
		keys = append(keys, character)
	}

	if sortKeys {
		sort.Slice(keys, func(firstIndex, secondIndex int) bool {
			return keys[firstIndex] < keys[secondIndex]
		})
	}

	return keys
}

File: src/typescript/helpers/data-structures/trie.ts
import { TrieNode } from "./trie-node";

type TraverseOptions = {
  sorted?: boolean; // if true, visit words in lexicographic order
};

export class Trie {
  root = new TrieNode();

  insert(word: string): void {
    let currentNode = this.root;

    for (const character of word) {
      if (!currentNode.children.has(character)) {
        currentNode.children.set(character, new TrieNode());
      }
      currentNode = currentNode.children.get(character)!;
    }

    currentNode.isEndOfWord = true;
  }

  search(word: string): boolean {
    const node = this.findNode(word);
    return node !== null && node.isEndOfWord;
  }

  startsWith(prefix: string): boolean {
    return this.findNode(prefix) !== null;
  }

  autocomplete(prefix: string, options: TraverseOptions = {}): string[] {
    const startNode = this.findNode(prefix);
    if (startNode === null) {
      return [];
    }

    const words: string[] = [];
    this.collectWords(startNode, prefix, words, options);
    return words;
  }

  getAllWords(options: TraverseOptions = {}): string[] {
    const words: string[] = [];
    this.collectWords(this.root, "", words, options);
    return words;
  }

  countWords(): number {
    return this.countWordsFrom(this.root);
  }

  traverse(
    callback: (word: string) => void,
    options: TraverseOptions = {}
  ): void {
    this.traverseFrom(this.root, "", callback, options);
  }

  private findNode(prefix: string): TrieNode | null {
    let currentNode = this.root;

    for (const character of prefix) {
      const nextNode = currentNode.children.get(character);
      if (!nextNode) return null;
      currentNode = nextNode;
    }

    return currentNode;
  }

  private collectWords(
    node: TrieNode,
    prefix: string,
    words: string[],
    options: TraverseOptions
  ): void {
    this.traverseFrom(node, prefix, (word) => words.push(word), options);
  }

  private traverseFrom(
    node: TrieNode,
    prefix: string,
    visit: (word: string) => void,
    options: TraverseOptions
  ): void {
    if (node.isEndOfWord) {
      visit(prefix);
    }

    const childEntries = options.sorted
      ? [...node.children.entries()].sort(([a], [b]) => a.localeCompare(b))
      : node.children.entries();

    for (const [character, childNode] of childEntries) {
      this.traverseFrom(childNode, prefix + character, visit, options);
    }
  }

  private countWordsFrom(node: TrieNode): number {
    let count = node.isEndOfWord ? 1 : 0;

    for (const childNode of node.children.values()) {
      count += this.countWordsFrom(childNode);
    }

    return count;
  }
}
